lang/compiler.go

```diff
@@ -169,9 +169,14 @@ type FluxCompiler struct {
 }
 
 func wrapFileJSONInPkg(bs []byte) []byte {
+	source := string(bs)
+	pkgHeader := `{"type":"Package"`
+	if source[0:len(pkgHeader)] == pkgHeader {
```

Comment:
I'm not sure what's going on here. Can you add a comment? It seems to be checking for a specific string but I'm a bit concerned that the look for an exact match will create problems. There's no requirement that the JSON start with the type.

Reply:
When the UI pass in the `extern` JSON, it was always a `File`, to enable it to be merged with the user's query, we used this function to wrap it into a package. But our compiler always generates a `Package`. When I need to generate some `extern` JSON objects in tests using our compiler, it will cause problems here.

There is no requirement that the JSON start with the type, but if it has type and the type is already a package, we do not wrap it again. I feel it is safe to do so.

Reply:
This still makes me very uncomfortable. Is there a way that we can unwrap the package from the compiler so that we can pass a file in here? Or I wonder if a separate field in the struct would work.

```go
type ASTCompiler struct {
    Extern json.RawMessage `json:"extern,omitempty"`
    ExternPackage flux.ASTHandle `json:"-"`
    AST    json.RawMessage `json:"ast"`
    Now    time.Time
}
```

This way you can pass in the ast handle directly and there's no need to serialize and then deserialize again.

Reply:
I actually wanted to change UI's template for generating the extern API, so they can directly generate a package, not waiting for us to add this hard-coded wrapper: https://github.com/influxdata/influxdb/blob/master/ui/src/variables/utils/buildVarsOption.ts#L9

Reply:
I don't think I'm comfortable with that. This would be a backwards incompatible change and it makes me more uncomfortable to attempt to do inspection of the JSON to see if we are wrapping a package or a file. It's relying on implementation details rather than an established way to perform JSON introspection.

It may be true that the extern should have likely been a package but this is an external API and I would not be comfortable changing this at this point. If we want to change the structure of the AST compiler, I think we need to create a version 2 or add an additional JSON attribute to the existing compiler.

If we want to change that or overload this operator I think that needs to be part of a separate issue and discussion. Adding an additional struct field that takes an AST handle directly mitigates the backwards compatibility aspects and solves the immediate problem if we want to go with the code snippet I have above.

Reply:
I handled this problem by tweaking the test code in order to avoid complicating the product code.